File: java/src/prep/leetcode/LongestRepeatingCharacterReplacement.java
package prep.leetcode;

public class LongestRepeatingCharacterReplacement 
{
	public static int characterReplacement(String s, int k) 
	{
		int maxLength=0;
		int maxCount=0;
		int left=0;
		int[] charCounts=new int[26];
		
		for(int right=0;right<s.length();right++)
		{
			int currentCharInd = s.charAt(right)-'A';
			charCounts[currentCharInd]++;
			maxCount=Math.max(maxCount, charCounts[currentCharInd]);
			
			while(right-left-maxCount+1 > k)
			{
				charCounts[s.charAt(left)-'A']--;
				left++;
			}
			maxLength=Math.max(maxLength, (right-left)+1);
		}
		
		return maxLength;
        
    }

	public static void main(String[] args)
	{
		System.out.println(characterReplacement("AABABBA",1));

	}

}

File: python/twopointer/LongestRepeatingCharacterReplacement.py
def characterReplacement(self, s: str, k: int) -> int:
    maxLength = 0
    char_count = {}
    left = 0

    for right in range(len(s)):
        char_count[s[right]] = char_count.get(s[right], 0) + 1
        while (right - left + 1) - max(char_count.values()) > k:
            char_count[s[left]] = char_count.get(s[left]) - 1
            left += 1
        maxLength = max(maxLength, right - left + 1)
    return maxLength